controllers/configurationpolicy_controller.go

```diff
@@ -940,378 +1106,250 @@ func (r *ConfigurationPolicyReconciler) handleObjectTemplates(plc *policyv1.Conf
-
-			resolveOptions.Watcher = &objID
-		} else {
-			tmplResolver, err = templates.NewResolver(r.TargetK8sConfig, templates.Config{})
-		}
-
-		if err != nil {
-			return err
-		}
-
-		var objTemps []*policyv1.ObjectTemplate
-
-		// process object templates for go template usage
-		for i, rawData := range rawDataList {
-			if templates.HasTemplate(rawData, "", true) {
-				log.V(1).Info("Processing policy templates")
-
-				// If there's a template, we can't rely on the cache results.
-				r.processedPolicyCache.Delete(plc.GetUID())
-
-				resolvedTemplate, tplErr := tmplResolver.ResolveTemplate(rawData, nil, &resolveOptions)
-
-				// If the error is because the padding is invalid, this either means the encrypted value was not
-				// generated by the "protect" template function or the AES key is incorrect. Control for a stale
-				// cached key.
-				if usedKeyCache && (errors.Is(tplErr, templates.ErrInvalidPKCS7Padding) ||
-					errors.Is(tplErr, templates.ErrInvalidAESKey) ||
-					errors.Is(tplErr, templates.ErrAESKeyNotSet)) {
-					log.V(2).Info(
-						"The template decryption failed likely due to an invalid encryption key, will refresh " +
-							"the encryption key cache and try the decryption again",
-					)
-					var encryptionConfig templates.EncryptionConfig
-					var err error
-
-					encryptionConfig, usedKeyCache, err = r.getEncryptionConfig(plc, true)
-					if err != nil {
-						addTemplateErrorViolation("", err.Error())
-
-						return err
-					}
-
-					resolveOptions.EncryptionConfig = encryptionConfig
+	var tmplResolver *templates.TemplateResolver
 
-					resolvedTemplate, tplErr = tmplResolver.ResolveTemplate(rawData, nil, &resolveOptions)
-				}
+	if usingWatch {
+		tmplResolver, err = templates.NewResolverWithDynamicWatcher(r.DynamicWatcher, templates.Config{})
+		objID := plc.ObjectIdentifier()
 
-				if tplErr != nil {
-					var msg string
-					var returnedErr error
+		resolveOptions.Watcher = &objID
+	} else {
+		tmplResolver, err = templates.NewResolver(r.TargetK8sConfig, templates.Config{})
+	}
 
-					if errors.Is(tplErr, templates.ErrInvalidAESKey) || errors.Is(tplErr, templates.ErrAESKeyNotSet) {
-						msg = `The "policy-encryption-key" Secret contains an invalid AES key`
-						returnedErr = tplErr
-					} else if errors.Is(tplErr, templates.ErrInvalidIV) {
-						msg = fmt.Sprintf(
-							`The "%s" annotation value is not a valid initialization vector`, IVAnnotation,
-						)
+	if err != nil {
+		return parentStatusUpdateNeeded, err
+	}
 
-						returnedErr = fmt.Errorf("%w: %w", ErrPolicyInvalid, tplErr)
-					} else {
-						msg = tplErr.Error()
-					}
+	var objTemps []*policyv1.ObjectTemplate
 
-					addTemplateErrorViolation("", msg)
+	// process object templates for go template usage
+	for i, rawData := range rawDataList {
+		if !templates.HasTemplate(rawData, "", true) && !isRawObjTemplate {
+			continue
+		}
 
-					return returnedErr
-				}
+		if !templates.HasTemplate(rawData, "", true) {
```

Comment:
I'd like to avoid calling `HasTemplate` a second time since this can be expensive as it scans the whole string.